Project Proposal: Document all media properties (#855)
WordPress/openverse: rfcs/document_all_media_properties/20230307-project_proposal.md

## Assigned reviewers

- [ ] @AetherUnbound
- [ ] @krysal

## Description

### Year Goal

Overcoming **Data Inertia**.

### Motivation

The first step in making the data we have reliable is describing it in detail.
Currently, we have more than 600 million rows of data about media in the catalog
database and similar but slightly different sets of data in the API database and
elasticsearch. Because of the history of the project, some data we have does not
conform to the requirements we now set and does not have the fields we describe
as required, and there are some fields that store a very different data piece
than what the users might expect[^1].

The main goal of this project is to establish a comprehensive baseline of our
media data: what information we currently have about each media item and what we
expect to have. Our database currently holds more than 700 million rows of media
data. The same data is in our API database and Elasticsearch, but the data
properties there are slightly different. Due to the project's history, some data
in the catalog does not meet our current requirements and lacks necessary
fields, while other fields store data that doesn't align with user

> expectations[^1].

Reviewer: This looks to be a rephrasing of the prior paragraph?

Author: Oops, you're right! I'll remove the first paragraph.

By documenting all data properties, we can ensure that we have a complete
understanding of the data. This documentation will improve the API user and the
Openverse maintainer experience. It will also be the basis for data
normalization which will lead to faster searches and a better experience for
catalog and API contributors, as well as API users.

### Success criteria

The project will be considered successful when the document described below is
created, reviewed by at least two maintainers, and published on the Openverse
documentation site.

### Implementors

> @rwidom - creator of the

Author: Thank you, @rwidom! I just wanted to clarify that I added you as an implementor because you've already done such a huge amount of this work in your table.  I am really happy for your enthusiasm, and will be happy to collaborate, but please, don't feel that you _have_ to work more on this :) 

Let us see what we can do with regards to the database access.

> @rwidom - creator of the

Reviewer 2: Sounds great! Thanks again for pushing it forward @obulat !

[image table](https://docs.google.com/spreadsheets/d/1gaVsvFnsYby2iwzRm0Ta9IPhvNixCpQdHByGLo1r_pg/edit?usp=sharing)
("image table ddl" tab)

@obulat - main implementor @openverse-team - reviewers

### Implementation details

As this is mainly a documentation project, its implementation plan can be
written here, inside the project plan. (I think splitting it will unnecessarily
slow down a smallish project. Please, correct me if you think I'm wrong) This
project will create a table with all the media properties that we store or
compute on all levels of the stack (Catalog, API database, API elasticsearch,
Frontend). This table will serve as a basis for any work on data normalization
work. As such, it will be a work in progress because the links to the code and
the details about the properties should change as we progress through the data
normalization project. We can publish it on the monorepo documentation site or
the Openverse blog on Make WordPress, with the admonition that the table is a

> living document and will be updated.

Reviewer 2: I don't know if it makes sense for this proposal, but I can imagine a world in which the provider data ingester has a required method for each of the required fields, and the spreadsheet is populated by the doc strings for those methods, and then there is some CI like what we have for the DAG doc strings. It would be different for SQL based providers, but those methods could return part of a SQL statement. It would be a big refactor to get from our world to that world, but if the goal is to move data normalization upstream, we might be able to get there... ???

> living document and will be updated.

Reviewer: That could work for the catalog end perhaps! I was mostly thinking of a script that would go check the DDLs for our tables and ensure that each field is represented in the table (and vice versa). Just something very simple and straightforward for the cope of this effort. We could get more complicated down the road, but even a check to make sure that the [catalog DDL fields](https://github.com/WordPress/openverse-catalog/blob/b173d6e2dbf72b41728baed18ee8334d862b6be8/docker/local_postgres/0003_openledger_image_schema.sql#L16-L42) were present in the documentation and same for the [API model](https://github.com/WordPress/openverse/blob/main/api/catalog/api/models/image.py#L43) (though the latter would require something to actually load the `Image` model and access properties rather than just read lines in a file, which would be sufficient on the catalog end... 🤔).

The table will start with a list of columns in the Catalog database. For each
property, we will provide the following information:

- A clear definition of what the property represents.
- An explanation of where the property is sourced from in the catalog, and any

>   special considerations that should be taken into account when extracting it.

Author: That's a great point, @rwidom! I think this _will_ be different for different fields. Data set by Openverse (like the `created_on`) should be easier to describe. For the data parsed from the providers, I envisaged something like this:

### title
Extracted by the provider script, `title` in the `add_item` method([1](https://github.com/WordPress/openverse-catalog/blob/370b916f948d43791b07e1442d82e3c377142818/openverse_catalog/dags/providers/provider_api_scripts/provider_data_ingester.py#L423) or [2](https://github.com/WordPress/openverse-catalog/blob/370b916f948d43791b07e1442d82e3c377142818/openverse_catalog/dags/common/storage/image.py#L44)). Some providers, mainly the ones that collect user-submitted items, can have no title. This is represented by a blank string `""` in the catalog database. Some titles from the older media can have encoding issues[^1]. This is [mitigated in the frontend](https://github.com/WordPress/openverse/blob/90109616cee1303acd01741ade1fb4698528c584/frontend/src/utils/decode-data.ts#L8).

[^1]: ASCII/Unicode characters with escape sequences have escaped backslashes, which prevents them from rendering

>   special considerations that should be taken into account when extracting it.

Reviewer 2: Oh wow, that will be a huge improvement. Awesome.

  For instance, in the case of the `creator` field, we will specify whether to
  use the `author`, `photographer`, or both.
- The property's data type, and whether it is required or nullable. We will also
  indicate if the property is required but not present in some data, such as
  `license_url` in the `meta_data` field. If the property has a set of distinct
  values, we will note where they are in the code.
- Whether the property is served by the API, and if so, on which endpoints
  (search or single result, or both). We will also highlight any differences
  between the property's type in the API versus the catalog, as noted in the
  database, Elasticsearch, or API documentation.

> - If the API does not take the data directly from the catalog database, we will

Reviewer 2: OMG, I can't tell you how happy this makes me!  This will be so awesome!

  indicate how the data is computed, and where this occurs (during data refresh,
  indexing, or as a fallback in the API). We will provide a link to the relevant
  code, if available.

> - We will also document whether the property is used by the frontend, whether

Reviewer 2: !!!!!!

  the type matches the catalog/API response type, and whether the frontend
  performs any cleanup or sanitization of the property. A link to the relevant
  code will be provided, if available.

#### Tools

The main tool to use for this table is an online spreadsheet app. I'm a little
uncertain of whether the amount of information would be too much to fit into one
table. If it is, it might be a good idea to split the data into a table with
more standard pieces of information, and a list of all data properties with
longer descriptions and things like code links and the criteria for selecting
data from provider APIs.

#### The process for updating the table

In the future, when the data is cleaned or changed in any way, we will need to
update the table accordingly. If we keep it in the documentation site, we can
use PRs to keep the history of changes, linking to all the related PRs.

#### Localization concerns

We should identify whether any pieces of data are translated or left as is, and
whether it might be necessary (and possible) to translate any pieces. There is
an [open issue](https://github.com/WordPress/openverse/issues/589) about
translating audio genres, for example.

> 

Author: ### Other concerns

It will not be possible to document all of the fields inside the `meta_data` column. It is a JSON column that was used to collect "whatever doesn't fit in the other fields" and its values can vary greatly. It would be extremely time- and resources- consuming to try to extract _all_available `meta_data` fields. For this column, we will describe what we currently expect as a bare minimum value for it, and also the meta_data fields that are used in the provider scripts
```suggestion

```

> 

Author: Other concerns

It will not be possible to document all of the fields inside the meta_data column. It is a JSON column that was used to collect "whatever doesn't fit in the other fields" and its values can vary greatly. It would be extremely time- and resources- consuming to try to extract _all_available meta_data fields. For this column, we will describe what we currently expect as a bare minimum value for it, and also the meta_data fields that are used in the provider scripts

Reviewer 2: Does the data that doesn't fit anywhere else ever get used by the API or front end? I know that this is probably beyond the scope of this proposal, but I feel like unless we know that the API and/or front end are using those fields, we could keep them in S3 or something for a rainy day, but keeping them in postgres just degrades performance. :/

Author: As far as I know, we only use the `license_url` and `description` fields from the `meta_data` column. I am not sure how much other data we have in there, so creating this table will be a great opportunity to start exploring :) I don't think I can comment on keeping the data in S3 or the performance degradation until I have more data, but I would love to find out.

documentation site.

### Implementors

@rwidom - creator of the
[image table](https://docs.google.com/spreadsheets/d/1gaVsvFnsYby2iwzRm0Ta9IPhvNixCpQdHByGLo1r_pg/edit?usp=sharing)
("image table ddl" tab)

@obulat - main implementor @openverse-team - reviewers

### Implementation details

As this is mainly a documentation project, its implementation plan can be
written here, inside the project plan. (I think splitting it will unnecessarily
slow down a smallish project. Please, correct me if you think I'm wrong) This
project will create a **table** with all the media properties that we store or
compute on all levels of the stack (Catalog, API database, API elasticsearch,
Frontend).

#### Table as a living document

This table will serve as a basis for any work on data normalization work. As
such, it will be a work in progress because the links to the code and the
details about the properties should change as we progress through the data
normalization project. To ensure that the table is up-to-date, this project will
also add automation to the catalog CI pipeline to detect changes to the fields
in the table and the catalog DDL and add a comment to update them, similar to
the

> [migration safety warning](https://github.com/WordPress/openverse/blob/main/.github/workflows/migration_safety_warning.yml)

Reviewer: Agreed, while this would be ideal to run GitHub for access to the repo code, it would not be ideal if we needed to access prod resources as part of it. Since we already have all the connection processes set up for Airflow, I think a DAG sounds great for this purpose!

more standard pieces of information, and a list of all data properties with
longer descriptions and things like code links and the criteria for selecting
data from provider APIs.

#### The process for updating the table

In the future, when the data is cleaned or changed in any way, we will need to
update the table accordingly. If we keep it in the documentation site, we can
use PRs to keep the history of changes, linking to all the related PRs.

#### Localization concerns

We should identify whether any pieces of data are translated or left as is, and
whether it might be necessary (and possible) to translate any pieces. There is
an [open issue](https://github.com/WordPress/openverse/issues/589) about
translating audio genres, for example.

#### Limitations for `meta_data` field

It will not be possible to document all the fields inside the `meta_data`
column. It is a JSON column that was used to collect "whatever doesn't fit in
the other fields" and its values can vary greatly. It would be extremely time-
and resources- consuming to try to extract \_all_available meta_data fields. For
this column, we will describe what we currently expect as a bare minimum value

> for it, and also the meta_data fields that are used in the provider scripts

Reviewer 3: This is unfortunate and I think we should correct going forward giving recommendation of what to store here. In my mind it was designated for the popularity metrics.

As this is mainly a documentation project, its implementation plan can be
written here, inside the project plan. (I think splitting it will unnecessarily
slow down a smallish project. Please, correct me if you think I'm wrong) This
project will create a **table** with all the media properties that we store or
compute on all levels of the stack (Catalog, API database, API elasticsearch,
Frontend).

#### Table as a living document

This table will serve as a basis for any work on data normalization work. As
such, it will be a work in progress because the links to the code and the
details about the properties should change as we progress through the data
normalization project. To ensure that the table is up-to-date, this project will
also add automation to the catalog CI pipeline to detect changes to the fields
in the table and the catalog DDL and add a comment to update them, similar to
the
[migration safety warning](https://github.com/WordPress/openverse/blob/main/.github/workflows/migration_safety_warning.yml)

We will publish the table on the

> [Openverse documentation site](https://wordpress.github.io/openverse/), with the

Reviewer: Since this changed recently 😄 
```suggestion
[Openverse documentation site](https://docs.openverse.org/), with the
```

an [open issue](https://github.com/WordPress/openverse/issues/589) about
translating audio genres, for example.

#### Limitations for `meta_data` field

It will not be possible to document all the fields inside the `meta_data`
column. It is a JSON column that was used to collect "whatever doesn't fit in
the other fields" and its values can vary greatly. It would be extremely time-

> and resources- consuming to try to extract \_all_available meta_data fields. For

Reviewer: Should this be?
```suggestion
and resource-consuming to try to extract _all_ available meta\_data fields. For
```
